### PR TITLE
DAOS-7219 control: trim ssd details in basic storage scan

### PR DESCRIPTION
**Pull request by tanabarr**
When dmg storage scan is run without options, summary should group hosts
with similar storage configurations even if SSD details such as model
and firmware version differ. NVMe namespace capacity differences should
be reflected in summary output. Update hashstructure module to v2.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/cmd/dmg/storage_test.go [hunk 1]
```diff
@@ -71,6 +71,12 @@ func TestStorageCommands(t *testing.T) {
 			printRequest(t, &control.StorageScanReq{NvmeMeta: true}),
 			nil,
 		},
+		{
+			"Scan NVMe health with verbose",
+			"storage scan --nvme-health --verbose",
+			"",
+			errors.New("cannot use --verbose"),
+		},
```
Comment (kjacque) [inline src/control/cmd/dmg/storage_test.go:79]: Should also test the combination with `--nvme-meta` and without the conflicting options.

**Review by kjacque — COMMENTED** [review]
As I'm looking at this, and refreshing myself on the context, I'm wondering about the reasoning for passing the NvmeBasic field all the way down to the server level, and using it to prune what the server sends back, instead of making the decisions about what to print entirely at the dmg level. Are there observed/anticipated problems with storage scans using too much network bandwidth at scale? I'm sure we must have gone over this before but I could use a refresher. Thanks.

**Comment by tanabarr** [comment]
> As I'm looking at this, and refreshing myself on the context, I'm wondering about the reasoning for passing the NvmeBasic field all the way down to the server level, and using it to prune what the server sends back, instead of making the decisions about what to print entirely at the dmg level. Are there observed/anticipated problems with storage scans using too much network bandwidth at scale? I'm sure we must have gone over this before but I could use a refresher. Thanks.

The HostStorageMap creates hostset groupings based on the hash of individual response HostStorage contents, the NvmeDevices field of that struct should not contain SSD details at the point that it is added to the map (addResponse) so that hosts with identical PCI configs (but different SSD models for example) get grouped together. This might be possible at the control API layer but is not possible in dmg. I think it's cleaner filtering in the server rather than manipulating the response in the control API but I'm happy to change if there is a consensus.

**Comment by kjacque** [comment]
> The HostStorageMap creates hostset groupings based on the hash of individual response HostStorage contents, the NvmeDevices field of that struct should not contain SSD details at the point that it is added to the map (addResponse) so that hosts with identical PCI configs (but different SSD models for example) get grouped together. This might be possible at the control API layer but is not possible in dmg. Thinking about it I would prefer to keep the filtering server side as it looks a bit awkward being placed in the control API.

Hm, okay. It still feels a little too invasive to me to pass it all the way down to the server, in the interest of constructing the response specifically based on how the tool displays it. In my view it would be better for the higher layer that makes the map to worry about how to organize it. You could use different functions to choose the map keys based on that NvmeBasic parameter, maybe.

**Comment by tanabarr** [comment]
> > The HostStorageMap creates hostset groupings based on the hash of individual response HostStorage contents, the NvmeDevices field of that struct should not contain SSD details at the point that it is added to the map (addResponse) so that hosts with identical PCI configs (but different SSD models for example) get grouped together. This might be possible at the control API layer but is not possible in dmg. Thinking about it I would prefer to keep the filtering server side as it looks a bit awkward being placed in the control API.
> 
> Hm, okay. It still feels a little too invasive to me to pass it all the way down to the server, in the interest of constructing the response specifically based on how the tool displays it. In my view it would be better for the higher layer that makes the map to worry about how to organize it. You could use different functions to choose the map keys based on that NvmeBasic parameter, maybe.

I'm struggling to make progress with this and https://github.com/daos-stack/daos/pull/5350 because of review pushback, I'm trying to implement feature changes requested whilst make the code robust and on one hand I'm being asked not to make large changes and to keep alterations minimal but on the other hand I'm being asked to alter existing design. I looked at implementing the filtering within the control API but anticipated pushback from @mjmac based on adding context specific details into what could be considered context agnostic request/response processing. @mjmac let me know if you want me to move the filtering into StorageScanResp in the control API.

**Comment by tanabarr** [comment]
Example output with this change (results grouped even if SSD model differs), SSD capacity differences will result in separate summary entries:
```
bash-4.2$ dmg storage scan
Hosts          SCM Total             NVMe Total
-----          ---------             ----------
wolf-[130-133] 6.4 TB (2 namespaces) 4.7 TB (4 controllers)
bash-4.2$ dmg storage scan --verbose
--------
wolf-130
--------
SCM Namespace Socket ID Capacity
------------- --------- --------
pmem0         0         3.2 TB
pmem1         1         3.2 TB

NVMe PCI     Model               FW Revision Socket ID Capacity
--------     -----               ----------- --------- --------
0000:5e:00.0 INTEL SSDPE2KE016T8 VDV10170    0         1.6 TB
0000:5f:00.0 INTEL SSDPE2KE016T8 VDV10170    0         1.6 TB
0000:81:00.0 INTEL SSDPED1K750GA E2010475    1         750 GB
0000:da:00.0 INTEL SSDPED1K750GA E2010475    1         750 GB

--------------
wolf-[131-133]
--------------
SCM Namespace Socket ID Capacity
------------- --------- --------
pmem0         0         3.2 TB
pmem1         1         3.2 TB

NVMe PCI     Model               FW Revision Socket ID Capacity
--------     -----               ----------- --------- --------
0000:5e:00.0 INTEL SSDPE2KE016T8 VDV10170    0         1.6 TB
0000:5f:00.0 INTEL SSDPE2KE016T8 VDV10170    0         1.6 TB
0000:81:00.0 INTEL SSDPED1K750GA E2010435    1         750 GB
0000:da:00.0 INTEL SSDPED1K750GA E2010435    1         750 GB

```

**Comment by kjacque** [comment]
> I'm trying to implement feature changes requested whilst make the code robust and on one hand I'm being asked not to make large changes and to keep alterations minimal but on the other hand I'm being asked to alter existing design.

Understood. The only reason I bring it up is because it looks more like new work, so maybe easier to make the change. It's not a huge deal, it just seems counterintuitive to me to let display options bleed so far into the call hierarchy. Perfectly willing to disagree and proceed if we just need to get it done, but I imagine it will be harder to clean it up later if that's desired.

**Comment by tanabarr** [comment]
> > I'm trying to implement feature changes requested whilst make the code robust and on one hand I'm being asked not to make large changes and to keep alterations minimal but on the other hand I'm being asked to alter existing design.
> 
> Understood. The only reason I bring it up is because it looks more like new work, so maybe easier to make the change. It's not a huge deal, it just seems counterintuitive to me to let display options bleed so far into the call hierarchy. Perfectly willing to disagree and proceed if we just need to get it done, but I imagine it will be harder to clean it up later if that's desired.

NvmeBasic was implemented as requested by Johann previously during the config generate command work, this just reuses it in storage scan.

**Review by mjmac — COMMENTED** [review]
Doing some filtering on the server side makes sense because there is additional expense to get things like the health stats. To Kris' point though, it does seem kind of strange to have a special RPC flag that results in getting all of the controller information and then throwing most of it away on the server side. That is basically presentation logic running in the RPC handler, which is not great because it means that we are more likely to have to change stuff on the server in order to support a change in presentation (e.g. what if we add a new feature that wants information somewhere between the normal set of info and what NvmeBasic returns? Do we add yet another flag to the RPC?). Missed that detail when looking at the autoconfig PR when it landed.

We could maybe justify keeping the flag as a performance optimization if it's possible to use that flag in order to avoid getting the information in the first place, similar to the way we avoid getting all of the health stats unless they've been requested. If that's not practical, I do think that NvmeBasic should be considered technical debt in need of refactoring to reduce complexity in the RPC payloads and handlers. It doesn't necessarily have to happen as part of this PR though.

**Comment by tanabarr** [comment]
> Doing some filtering on the server side makes sense because there is additional expense to get things like the health stats. To Kris' point though, it does seem kind of strange to have a special RPC flag that results in getting all of the controller information and then throwing most of it away on the server side. That is basically presentation logic running in the RPC handler, which is not great because it means that we are more likely to have to change stuff on the server in order to support a change in presentation (e.g. what if we add a new feature that wants information somewhere between the normal set of info and what NvmeBasic returns? Do we add yet another flag to the RPC?). Missed that detail when looking at the autoconfig PR when it landed.
> 
> We could maybe justify keeping the flag as a performance optimization if it's possible to use that flag in order to avoid getting the information in the first place, similar to the way we avoid getting all of the health stats unless they've been requested. If that's not practical, I do think that NvmeBasic should be considered technical debt in need of refactoring to reduce complexity in the RPC payloads and handlers. It doesn't necessarily have to happen as part of this PR though.

Because in this case the filtering is just masking certain fields within a struct that will normally be returned from cached scan results, there isn't much to be gained in terms of performance by applying the filter. I'm okay with moving the field masking into the control API if you feel that would be more appropriate but I'm not sure what that buys us and to me it doesn't really look like a good place to be applying the filter given it adds implicit knowledge of the response contents to an area of code that doesn't otherwise need to understand the details.

**Comment by mjmac** [comment]
> 
> Because in this case the filtering is just masking certain fields within a struct that will normally be returned from cached scan results, there isn't much to be gained in terms of performance by applying the filter. I'm okay with moving the field masking into the control API if you feel that would be more appropriate but I'm not sure what that buys us and to me it doesn't really look like a good place to be applying the filter given it adds implicit knowledge of the response contents to an area of code that doesn't otherwise need to understand the details.

I don't follow this last point, please elaborate. Isn't deciding what information to display inherently a responsibility of the display logic?

**Comment by tanabarr** [comment]
> > Because in this case the filtering is just masking certain fields within a struct that will normally be returned from cached scan results, there isn't much to be gained in terms of performance by applying the filter. I'm okay with moving the field masking into the control API if you feel that would be more appropriate but I'm not sure what that buys us and to me it doesn't really look like a good place to be applying the filter given it adds implicit knowledge of the response contents to an area of code that doesn't otherwise need to understand the details.
> 
> I don't follow this last point, please elaborate. Isn't deciding what information to display inherently a responsibility of the display logic?

I don't consider the control API "display logic", do you?

**Comment by mjmac** [comment]
> > 
> > I don't follow this last point, please elaborate. Isn't deciding what information to display inherently a responsibility of the display logic?
> 
> I don't consider the control API "display logic", do you?

I thought the entire point of this discussion was based on the observation that we should probably be making decisions about what information is displayed to the user in the presentation logic, i.e. the pretty printers. Otherwise we have to make updates all through the stack in order to effect presentation-level changes. We don't have a perfect/clean separation currently due to some historical artifacts that weren't completely addressed when the switchover to the Control API happened, but I consider that technical debt to be fixed rather than precedent to be followed.

The ideal that we should be moving toward is that the client/server RPCs should be presentation-agnostic. As I mentioned earlier, it may be suitable to implement some server-side filtering to avoid expensive operations if the results of those operations won't be used, but in general we should lean toward just returning all available information in request responses and let the caller decide what to do with it. This reduces the maintenance cost of code over time by making it easier to make changes in isolation and without needing to test/fix complex filtering logic in the data layers.

**Comment by tanabarr** [comment]
> > > I don't follow this last point, please elaborate. Isn't deciding what information to display inherently a responsibility of the display logic?
> > 
> > 
> > I don't consider the control API "display logic", do you?
> 
> I thought the entire point of this discussion was based on the observation that we should probably be making decisions about what information is displayed to the user in the presentation logic, i.e. the pretty printers. Otherwise we have to make updates all through the stack in order to effect presentation-level changes. We don't have a perfect/clean separation currently due to some historical artifacts that weren't completely addressed when the switchover to the Control API happened, but I consider that technical debt to be fixed rather than precedent to be followed.
> 
> The ideal that we should be moving toward is that the client/server RPCs should be presentation-agnostic. As I mentioned earlier, it may be suitable to implement some server-side filtering to avoid expensive operations if the results of those operations won't be used, but in general we should lean toward just returning all available information in request responses and let the caller decide what to do with it. This reduces the maintenance cost of code over time by making it easier to make changes in isolation and without needing to test/fix complex filtering logic in the data layers.

The pertinent detail is that HostStorageMap will add the response to the same hash key if the results are the same when received in the control API layer, if that happens then unpacking the HSM to regroup results based on a filter applied in the presentation layer really doesn't make sense to me. Is that clear?

**Review by mjmac — APPROVED** [review]

**Review by kjacque — APPROVED** [review]

**Pull request merged**